### PR TITLE
Improves catalog crawling performance by using multiprocessing

### PR DESCRIPTION
**Pull request by lukecampbell**

### CODE REVIEW

**Comment by lukecampbell** [comment]
@kwilcox @daf Could you guys take a look at this when you have a moment?

**Comment by lukecampbell** [comment]
It's not as a dramatic improvement as I would have hoped but it did reduce crawling time by 75% (assuming a computer with 4 CPUs). There's just too many XML documents in a thredds catalog. Most of the catalogs I've tested this on make anywhere from 200 to 2000 HTTP requests. Each one is unique.

**Comment by kwilcox** [comment]
I had a local branch that did this and I crashed many many many THREDDS servers using it (most likely when calculating dataset size). I vote that the default worker count be 1.

This would break logging to a file (which I and a few other people I know of use with thredds_crawler)... you would need to implement something like this: https://github.com/axiom-data-science/paegan/blob/master/paegan/logger/multi_process_logging.py

I didn't look too closely at the code... are you handling recursive THREDDS catalogs now?

**Comment by lukecampbell** [comment]
The only multiprocessing bit of this is the HTTP requests and making the leaf dataset objects. There's only one line where the logging would need to get fixed. I haven't tested it rigorously on THREDDS servers, if you know of one that has a recursive catalog, I'd like to verify.

File: thredds_crawler/crawl.py [hunk 1]
```diff
@@ -26,16 +28,32 @@ def emit(self, record):
...
+def request_xml(url):
+    '''
+    Returns an etree.XMLRoot object loaded from the url
+    :param str url: URL for the resource to load as an XML
+    '''
+    try:
+        r = requests.get(url)
+        return r.text.encode('utf-8')
+    except BaseException:
+        logger.error("Skipping %s (error parsing the XML)" % url)
```
Comment (lukecampbell) [inline thredds_crawler/crawl.py]: This logging statement would need to be fixed

**Comment by kwilcox** [comment]
@lukecampbell I'll merge this, but first you'll need to make the MARACOOS TDS server not return a 500:
http://tds.maracoos.org/thredds/ so I can run the tests.

**Comment by kwilcox** [comment]
Thanks

**Comment by kwilcox** [comment]
I merged here: https://github.com/asascience-open/thredds_crawler/pull/18

**Comment by lukecampbell** [comment]
lol thanks, I forgot about this branch. I still need to patch logging. 

As for tds.maracoos.org, there are A LOT of harvesters that keep bringing it down. We're going to try to isolate the worst offenders and start blocking them if they bring it down. IOOS Catalog is one of them, but it runs between 2AM and 9AM. I think it's a university that's harvesting... but every time we reach out to everyone we know they all say they aren't doing it.

Sorry for the service being down, we're working on it.

**Pull request closed without merging**